ic-agent/src/agent/mod.rs

```diff
@@ -250,7 +269,71 @@ impl Agent {
         serde_cbor::from_slice(&bytes).map_err(AgentError::InvalidCborData)
     }
 
-    async fn submit_endpoint(&self, request: AsyncContent) -> Result<RequestId, AgentError> {
+    async fn query_endpoint<A>(
+        &self,
+        effective_canister_id: Principal,
+        request: QueryContent,
+    ) -> Result<A, AgentError>
+    where
+        A: serde::de::DeserializeOwned,
+    {
+        let request_id = to_request_id(&request)?;
+        let msg = self.construct_message(&request_id);
+        let signature = self.identity.sign(&msg).map_err(AgentError::SigningError)?;
+
+        let envelope = Envelope {
+            content: request,
+            sender_pubkey: signature.public_key,
+            sender_sig: signature.signature,
+        };
+
+        let mut serialized_bytes = Vec::new();
+        let mut serializer = serde_cbor::Serializer::new(&mut serialized_bytes);
+        serializer.self_describe()?;
+        envelope.serialize(&mut serializer)?;
+
+        let bytes = self
+            .transport
+            .query(effective_canister_id, serialized_bytes)
+            .await?;
+        serde_cbor::from_slice(&bytes).map_err(AgentError::InvalidCborData)
+    }
+
+    async fn read_state_endpoint<A>(
+        &self,
+        effective_canister_id: Principal,
```

Comment:
You don't always have a canister id though :(

Reply:
(Just copying here what I said on slack)

I think it’s fine to expose this functionality only when an effective canister id is provided. After all, you need to indicate where you want to read the state! And current use cases for `read_state` (call status, controller, module_hash) are related to a specific canister.